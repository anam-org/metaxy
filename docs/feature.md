## Feature Version

Every feature has a **feature version** - hash of its complete specification:
- Feature key
- Field definitions and code versions
- Dependencies (feature-level and field-level)

Being more pedantic, the feature version is calculated by hashing together its field versions and the feature key. The field version is calculated by hashing together the field code version, key, and versions of its dependencies.

Feature and field versions are static and deterministic. They are available as class methods:

```python
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=1),
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

# Get feature version
print(VideoProcessing.feature_version())  # "a3f8b2c1"

# Change code_version
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=2),  # Changed!
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

print(VideoProcessing.feature_version())  # "d7e9f4a2" (different!)
```

Every metadata row includes a `feature_version` column:
```python
metadata = pl.DataFrame({
    "sample_id": [1, 2, 3],
    "data_version": [...],
    "feature_version": "a3f8b2c1",  # Auto-added by store
})
```

This enables:
- **Current vs historical**: Filter by `feature_version` to get specific versions
- **Migration tracking**: Know which rows need updating
- **Audit trail**: See what feature definition produced each row


## User-Defined Metadata Alignment

When feature definitions change, you may need custom logic to align metadata with upstream changes. Override `Feature.align_metadata_with_upstream`:

### Default Behavior (Inner Join)

```python
class MyFeature(Feature, spec=...):
    # Default behavior: inner join on sample_id
    # Only keeps samples present in ALL upstream features
    pass

# During migration:
# - Reads current metadata
# - Inner joins with upstream on sample_id
# - Only samples present in both current AND upstream survive
```

### Custom Alignment Examples

#### One-to-Many Mapping

```python
class VideoFrames(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "frames"]),
    deps=[FeatureDep(key=FeatureKey(["videos"]))],
    ...
)):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Each video produces 30 frames.

        When videos upstream change, regenerate all frame sample IDs.
        """
        video_samples = upstream_metadata["videos"]["sample_id"]

        # Generate frame sample IDs
        frames = []
        for video_id in video_samples:
            for frame_idx in range(30):
                frames.append({
                    "sample_id": f"{video_id}_frame_{frame_idx}",
                    "video_id": video_id,
                    "frame_idx": frame_idx,
                })

        return pl.DataFrame(frames)
```

#### Conditional Filtering

```python
class ProcessedVideos(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Only process videos longer than 10 seconds."""

        videos = upstream_metadata["videos"]

        # Filter by duration
        valid_videos = videos.filter(pl.col("duration") > 10)

        # Keep custom columns from current_metadata if available
        if len(current_metadata) > 0:
            return current_metadata.join(
                valid_videos.select(pl.col("sample_id")),
                on="sample_id",
                how="inner",  # Only keep samples that pass filter
            )
        else:
            return valid_videos.select(["sample_id", "duration", "path"])
```

#### Outer Join (Keep All Samples)

```python
class MergedFeature(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Keep union of all upstream samples (outer join)."""

        all_samples = set()
        for upstream_df in upstream_metadata.values():
            all_samples.update(upstream_df["sample_id"].to_list())

        return pl.DataFrame({"sample_id": sorted(all_samples)})
```
