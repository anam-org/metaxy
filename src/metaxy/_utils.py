import narwhals as nw
import polars as pl


def collect_to_polars(lazy_frame: nw.LazyFrame) -> pl.DataFrame:
    """Helper to collect a Narwhals LazyFrame and convert to Polars DataFrame.

    This handles all backend conversions (Polars, DuckDB/PyArrow, etc.) transparently.
    Used throughout tests for materializing query results.
    """
    return lazy_frame.collect().to_polars()
