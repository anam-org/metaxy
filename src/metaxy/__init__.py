from metaxy.models.container import ContainerDep, ContainerSpec, SpecialContainerDep
from metaxy.models.feature import Feature
from metaxy.models.feature_spec import FeatureDep, FeatureDepMetadata, FeatureSpec
from metaxy.models.types import ContainerKey, FeatureKey

__all__ = [
    "Feature",
    "FeatureDep",
    "FeatureDepMetadata",
    "FeatureSpec",
    "ContainerDep",
    "ContainerSpec",
    "SpecialContainerDep",
    "FeatureKey",
    "ContainerKey",
]
