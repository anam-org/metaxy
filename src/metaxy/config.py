"""Configuration system for Metaxy using pydantic-settings."""

from pathlib import Path
from typing import TYPE_CHECKING, Any, TypeVar, overload

try:
    import tomllib  # Python 3.11+
except ImportError:
    import tomli as tomllib  # Fallback for Python 3.10

from pydantic import Field as PydanticField
from pydantic_settings import (
    BaseSettings,
    PydanticBaseSettingsSource,
    SettingsConfigDict,
)

if TYPE_CHECKING:
    from metaxy.metadata_store.base import MetadataStore
    from metaxy.models.feature import FeatureRegistry

T = TypeVar("T")


def load_registry(registry_path: str) -> "FeatureRegistry":
    """Load registry from import path.

    Args:
        registry_path: Import path like "myapp.features:my_registry"

    Returns:
        FeatureRegistry instance

    Raises:
        ValueError: If path is invalid
        ImportError: If module or registry not found
        TypeError: If loaded object is not a FeatureRegistry
    """
    from metaxy.models.feature import FeatureRegistry

    if ":" not in registry_path:
        raise ValueError(
            f"Invalid registry path: {registry_path}. "
            f"Expected format: 'module.path:registry_name'"
        )

    module_path, registry_name = registry_path.split(":", 1)

    try:
        module = __import__(module_path, fromlist=[registry_name])
        registry = getattr(module, registry_name)
    except (ImportError, AttributeError) as e:
        raise ImportError(f"Failed to load registry from {registry_path}: {e}") from e

    if not isinstance(registry, FeatureRegistry):
        raise TypeError(
            f"Registry at {registry_path} is not a FeatureRegistry instance. "
            f"Got: {type(registry)}"
        )

    return registry


class TomlConfigSettingsSource(PydanticBaseSettingsSource):
    """Custom settings source for TOML configuration files.

    Auto-discovers configuration in this order:
    1. Explicit file path if provided
    2. metaxy.toml in current directory (preferred)
    3. pyproject.toml [tool.metaxy] section (fallback)
    4. No config (returns empty dict)
    """

    def __init__(self, settings_cls: type[BaseSettings], toml_file: Path | None = None):
        super().__init__(settings_cls)
        self.toml_file = toml_file or self._discover_config_file()
        self.toml_data = self._load_toml()

    def _discover_config_file(self) -> Path | None:
        """Auto-discover config file."""
        # Prefer metaxy.toml
        if Path("metaxy.toml").exists():
            return Path("metaxy.toml")

        # Fallback to pyproject.toml
        if Path("pyproject.toml").exists():
            return Path("pyproject.toml")

        return None

    def _load_toml(self) -> dict[str, Any]:
        """Load TOML file and extract metaxy config."""
        if self.toml_file is None:
            return {}

        with open(self.toml_file, "rb") as f:
            data = tomllib.load(f)

        # Extract [tool.metaxy] from pyproject.toml or root from metaxy.toml
        if self.toml_file.name == "pyproject.toml":
            return data.get("tool", {}).get("metaxy", {})
        else:
            return data

    def get_field_value(self, field: Any, field_name: str) -> tuple[Any, str, bool]:
        """Get field value from TOML data."""
        field_value = self.toml_data.get(field_name)
        return field_value, field_name, False

    def __call__(self) -> dict[str, Any]:
        """Return all settings from TOML."""
        return self.toml_data


class StoreConfig(BaseSettings):
    """Configuration for a single metadata store.

    Structure:
        type: Full import path to store class
        config: Dict of all configuration (including fallback_stores)

    Example:
        >>> config = StoreConfig(
        ...     type="metaxy_delta.DeltaMetadataStore",
        ...     config={
        ...         "table_uri": "s3://bucket/metadata",
        ...         "region": "us-west-2",
        ...         "fallback_stores": ["prod"],
        ...     }
        ... )
    """

    model_config = SettingsConfigDict(
        extra="forbid",  # Only type and config fields allowed
    )

    # Store class (full import path)
    type: str

    # Store configuration (all kwargs for __init__)
    # This includes fallback_stores, table_uri, db_path, storage_options, etc.
    config: dict[str, Any] = PydanticField(default_factory=dict)


class MetaxyConfig(BaseSettings):
    """Main Metaxy configuration.

    Loads from:
    1. TOML file (metaxy.toml or pyproject.toml [tool.metaxy])
    2. Environment variables (METAXY_*)
    3. Init arguments

    Priority: init > env vars > TOML

    Example:
        >>> # Auto-discover config
        >>> config = MetaxyConfig.load()
        >>>
        >>> # Get store instance
        >>> store = config.get_store("prod")
        >>>
        >>> # Override via env var
        >>> # METAXY_STORE=staging METAXY_REGISTRY=myapp.features:my_registry
        >>> config = MetaxyConfig.load()
        >>> store = config.get_store()  # Uses staging with custom registry
    """

    model_config = SettingsConfigDict(
        env_prefix="METAXY_",
        env_nested_delimiter="__",
    )

    # Store to use
    store: str = "dev"

    # Named store configurations
    stores: dict[str, StoreConfig] = PydanticField(default_factory=dict)

    # Migrations directory
    migrations_dir: str = "metaxy/migrations"

    # Entrypoints to load (list of module paths)
    entrypoints: list[str] = PydanticField(default_factory=list)

    @classmethod
    def settings_customise_sources(
        cls,
        settings_cls: type[BaseSettings],
        init_settings: PydanticBaseSettingsSource,
        env_settings: PydanticBaseSettingsSource,
        dotenv_settings: PydanticBaseSettingsSource,
        file_secret_settings: PydanticBaseSettingsSource,
    ) -> tuple[PydanticBaseSettingsSource, ...]:
        """Customize settings sources: init → env → TOML.

        Priority (first wins):
        1. Init arguments
        2. Environment variables
        3. TOML file
        """
        toml_settings = TomlConfigSettingsSource(settings_cls)
        return (init_settings, env_settings, toml_settings)

    @classmethod
    def load(
        cls, config_file: str | Path | None = None, *, search_parents: bool = True
    ) -> "MetaxyConfig":
        """Load config with auto-discovery and parent directory search.

        Args:
            config_file: Optional config file path (overrides auto-discovery)
            search_parents: Search parent directories for config file (default: True)

        Returns:
            Loaded config (TOML + env vars merged)

        Example:
            >>> # Auto-discover with parent search
            >>> config = MetaxyConfig.load()
            >>>
            >>> # Explicit file
            >>> config = MetaxyConfig.load("custom.toml")

            >>> # Auto-discover without parent search
            >>> config = MetaxyConfig.load(search_parents=False)
        """
        # Search for config file if not explicitly provided
        if config_file is None and search_parents:
            config_file = cls._discover_config_with_parents()

        # For explicit file, temporarily patch the TomlConfigSettingsSource
        # to use that file, then use normal instantiation
        # This ensures env vars still work

        if config_file:
            # Create a custom settings source class for this file
            toml_path = Path(config_file)

            class CustomTomlSource(TomlConfigSettingsSource):
                def __init__(self, settings_cls: type[BaseSettings]):
                    # Skip auto-discovery, use explicit file
                    super(TomlConfigSettingsSource, self).__init__(settings_cls)
                    self.toml_file = toml_path
                    self.toml_data = self._load_toml()

            # Customize sources to use custom TOML file
            original_method = cls.settings_customise_sources

            @classmethod  # type: ignore[misc]
            def custom_sources(
                cls_inner,
                settings_cls,
                init_settings,
                env_settings,
                dotenv_settings,
                file_secret_settings,
            ):
                toml_settings = CustomTomlSource(settings_cls)
                return (init_settings, env_settings, toml_settings)

            # Temporarily replace method
            cls.settings_customise_sources = custom_sources  # type: ignore[assignment]
            config = cls()
            cls.settings_customise_sources = original_method  # type: ignore[method-assign]
        else:
            # Use default sources (auto-discovery + env vars)
            config = cls()

        # Load entrypoints if configured (do this after config is created)
        if config.entrypoints:
            from metaxy.entrypoints import load_config_entrypoints

            load_config_entrypoints(config.entrypoints)

        return config

    @staticmethod
    def _discover_config_with_parents() -> Path | None:
        """Discover config file by searching current and parent directories.

        Searches for metaxy.toml or pyproject.toml in current directory,
        then iteratively searches parent directories.

        Returns:
            Path to config file if found, None otherwise
        """
        current = Path.cwd()

        while True:
            # Check for metaxy.toml (preferred)
            metaxy_toml = current / "metaxy.toml"
            if metaxy_toml.exists():
                return metaxy_toml

            # Check for pyproject.toml
            pyproject_toml = current / "pyproject.toml"
            if pyproject_toml.exists():
                return pyproject_toml

            # Move to parent
            parent = current.parent
            if parent == current:
                # Reached root
                break
            current = parent

        return None

    def get_store(
        self,
        name: str | None = None,
    ) -> "MetadataStore":
        """Instantiate metadata store by name.

        Args:
            name: Store name (uses config.store if None)

        Returns:
            Instantiated metadata store

        Raises:
            ValueError: If store name not found in config, or if fallback stores
                have different hash algorithms than the parent store
            ImportError: If store class cannot be imported

        Example:
            >>> config = MetaxyConfig.load()
            >>> store = config.get_store("prod")
            >>>
            >>> # Use default store
            >>> store = config.get_store()
        """
        from metaxy.data_versioning.hash_algorithms import HashAlgorithm

        if len(self.stores) == 0:
            raise ValueError(
                "No Metaxy stores available. They should be configured in metaxy.toml|pyproject.toml or via environment variables."
            )

        name = name or self.store

        if name not in self.stores:
            raise ValueError(
                f"Store '{name}' not found in config. "
                f"Available stores: {list(self.stores.keys())}"
            )

        store_config = self.stores[name]

        # Import store class
        store_class = self._import_class(store_config.type)

        # Extract configuration
        config_copy = store_config.config.copy()
        fallback_store_names = config_copy.pop("fallback_stores", [])

        # Get hash_algorithm from config (if specified) and convert to enum
        configured_hash_algorithm = config_copy.get("hash_algorithm")
        if configured_hash_algorithm is not None:
            # Convert string to enum if needed
            if isinstance(configured_hash_algorithm, str):
                configured_hash_algorithm = HashAlgorithm(configured_hash_algorithm)
                config_copy["hash_algorithm"] = configured_hash_algorithm
        else:
            # Use default
            configured_hash_algorithm = HashAlgorithm.XXHASH64
            config_copy["hash_algorithm"] = configured_hash_algorithm

        # Build fallback stores recursively
        fallback_stores = []
        for fallback_name in fallback_store_names:
            fallback_store = self.get_store(fallback_name)
            fallback_stores.append(fallback_store)

        # Instantiate store with config + fallback_stores
        store = store_class(
            fallback_stores=fallback_stores,
            **config_copy,
        )

        # Verify the store actually uses the hash algorithm we configured
        # (in case a store subclass overrides the default or ignores the parameter)
        if store.hash_algorithm != configured_hash_algorithm:
            raise ValueError(
                f"Store '{name}' ({store_class.__name__}) was configured with "
                f"hash_algorithm='{configured_hash_algorithm.value}' but is using "
                f"'{store.hash_algorithm.value}'. The store class may have overridden "
                f"the hash algorithm. All stores must use the same hash algorithm."
            )

        return store

    @staticmethod
    def _import_class(class_path: str) -> type:
        """Import class from module path.

        Args:
            class_path: Full import path like "metaxy.metadata_store.InMemoryMetadataStore"

        Returns:
            Imported class

        Raises:
            ImportError: If module or class not found
        """
        module_path, class_name = class_path.rsplit(".", 1)
        module = __import__(module_path, fromlist=[class_name])
        return getattr(module, class_name)

    @staticmethod
    @overload
    def _import_object(import_path: str, expected_type: type[T]) -> T: ...

    @staticmethod
    @overload
    def _import_object(import_path: str, expected_type: None = None) -> Any: ...

    @staticmethod
    def _import_object(
        import_path: str, expected_type: type[T] | None = None
    ) -> T | Any:
        """Import object from module:object path.

        Args:
            import_path: Import path like "module.path:object_name"
            expected_type: Optional type to validate against

        Returns:
            Imported object (typed as expected_type if provided)

        Raises:
            ValueError: If path format is invalid
            ImportError: If module or object not found
            TypeError: If object doesn't match expected_type
        """
        if ":" not in import_path:
            raise ValueError(
                f"Invalid import path: {import_path}. "
                f"Expected format: 'module.path:object_name'"
            )

        module_path, object_name = import_path.split(":", 1)

        try:
            module = __import__(module_path, fromlist=[object_name])
            obj = getattr(module, object_name)
        except (ImportError, AttributeError) as e:
            raise ImportError(f"Failed to load object from {import_path}: {e}") from e

        if expected_type is not None and not isinstance(obj, expected_type):
            type_name = getattr(expected_type, "__name__", str(expected_type))
            raise TypeError(
                f"Object at {import_path} is not a {type_name} instance. "
                f"Got: {type(obj)}"
            )

        return obj
