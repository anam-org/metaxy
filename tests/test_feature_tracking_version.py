"""Tests for feature_tracking_version that includes project in the hash.

The tracking version is used for system tables (feature_versions) to isolate
features from different projects in the same metadata store. It differs from
feature_version (used for data versioning) in that it includes the project name.

Key behaviors:
1. feature_tracking_version changes when project changes
2. feature_version does NOT change when project changes
3. Two identical features in different projects have different tracking versions
4. feature_tracking_version is deterministic for the same (feature, project) pair
"""

from __future__ import annotations

from syrupy.assertion import SnapshotAssertion

from metaxy import Feature, FeatureDep, FeatureKey, FeatureSpec, FieldKey, FieldSpec
from metaxy.config import MetaxyConfig
from metaxy.models.feature import FeatureGraph


def test_feature_tracking_version_includes_project(snapshot: SnapshotAssertion) -> None:
    """Test that feature_tracking_version changes when project changes."""
    # Create same feature in two different projects
    config_a = MetaxyConfig(project="project_a")
    graph_a = FeatureGraph()

    MetaxyConfig.set(config_a)

    with graph_a.use():

        class FeatureInA(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["test", "feature"]),
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Same feature in project_b
    config_b = MetaxyConfig(project="project_b")
    graph_b = FeatureGraph()

    MetaxyConfig.set(config_b)

    with graph_b.use():

        class FeatureInB(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["test", "feature"]),
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # The features should have different projects
    assert FeatureInA.project != FeatureInB.project
    assert FeatureInA.project == "project_a"
    assert FeatureInB.project == "project_b"

    # feature_tracking_version should be DIFFERENT (includes project)
    tracking_version_a = FeatureInA.feature_tracking_version()
    tracking_version_b = FeatureInB.feature_tracking_version()
    assert tracking_version_a != tracking_version_b

    # But feature_version should be SAME (data versioning unchanged by project)
    assert FeatureInA.feature_version() == FeatureInB.feature_version()

    # Snapshot for verification
    assert {
        "project_a": FeatureInA.project,
        "project_b": FeatureInB.project,
        "tracking_version_a": tracking_version_a,
        "tracking_version_b": tracking_version_b,
        "feature_version_a": FeatureInA.feature_version(),
        "feature_version_b": FeatureInB.feature_version(),
        "tracking_versions_differ": tracking_version_a != tracking_version_b,
        "feature_versions_same": FeatureInA.feature_version()
        == FeatureInB.feature_version(),
    } == snapshot

    MetaxyConfig.reset()


def test_feature_version_unchanged_by_project(snapshot: SnapshotAssertion) -> None:
    """Test that feature_version does NOT change when only project changes.

    This is critical: feature_version is used for data versioning and should
    only change when the computational definition changes, not metadata like project.
    """
    # Define identical feature specs in two projects
    graph_1 = FeatureGraph()
    graph_2 = FeatureGraph()

    config_1 = MetaxyConfig(project="project_1")
    MetaxyConfig.set(config_1)

    with graph_1.use():

        class Feature1(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["data", "feature"]),
                fields=[
                    FieldSpec(key=FieldKey(["frames"]), code_version=1),
                    FieldSpec(key=FieldKey(["audio"]), code_version=2),
                ],
            ),
        ):
            pass

    config_2 = MetaxyConfig(project="project_2")
    MetaxyConfig.set(config_2)

    with graph_2.use():

        class Feature2(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["data", "feature"]),
                fields=[
                    FieldSpec(key=FieldKey(["frames"]), code_version=1),
                    FieldSpec(key=FieldKey(["audio"]), code_version=2),
                ],
            ),
        ):
            pass

    # feature_version should be IDENTICAL (data versioning unchanged)
    version_1 = Feature1.feature_version()
    version_2 = Feature2.feature_version()

    assert version_1 == version_2
    assert version_1 == snapshot

    # But projects should differ
    assert Feature1.project != Feature2.project

    MetaxyConfig.reset()


def test_tracking_version_deterministic(snapshot: SnapshotAssertion) -> None:
    """Test that tracking version (project + feature) is deterministic."""
    config = MetaxyConfig(project="deterministic_project")
    MetaxyConfig.set(config)

    try:

        class TestFeature(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["test", "deterministic"]),
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

        # feature_tracking_version should be deterministic
        tracking_1 = TestFeature.feature_tracking_version()
        tracking_2 = TestFeature.feature_tracking_version()

        assert tracking_1 == tracking_2
        assert len(tracking_1) == 64  # SHA256 hex
        assert tracking_1 == snapshot

    finally:
        MetaxyConfig.reset()


def test_feature_with_deps_different_projects(snapshot: SnapshotAssertion) -> None:
    """Test feature_version stays same across projects even with dependencies."""
    # Project A: upstream and downstream
    config_a = MetaxyConfig(project="project_a")
    graph_a = FeatureGraph()

    MetaxyConfig.set(config_a)

    with graph_a.use():

        class UpstreamA(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["upstream"]),
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

        class DownstreamA(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["downstream"]),
                deps=[FeatureDep(feature=FeatureKey(["upstream"]))],
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # Project B: same features
    config_b = MetaxyConfig(project="project_b")
    graph_b = FeatureGraph()

    MetaxyConfig.set(config_b)

    with graph_b.use():

        class UpstreamB(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["upstream"]),
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

        class DownstreamB(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["downstream"]),
                deps=[FeatureDep(feature=FeatureKey(["upstream"]))],
                fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
            ),
        ):
            pass

    # feature_version should be same across projects
    assert UpstreamA.feature_version() == UpstreamB.feature_version()
    assert DownstreamA.feature_version() == DownstreamB.feature_version()

    # But projects should differ
    assert UpstreamA.project == "project_a"
    assert UpstreamB.project == "project_b"
    assert DownstreamA.project == "project_a"
    assert DownstreamB.project == "project_b"

    # Snapshot both
    assert {
        "upstream_a_version": UpstreamA.feature_version(),
        "upstream_b_version": UpstreamB.feature_version(),
        "downstream_a_version": DownstreamA.feature_version(),
        "downstream_b_version": DownstreamB.feature_version(),
        "upstream_a_project": UpstreamA.project,
        "upstream_b_project": UpstreamB.project,
        "downstream_a_project": DownstreamA.project,
        "downstream_b_project": DownstreamB.project,
    } == snapshot

    MetaxyConfig.reset()


def test_project_in_graph_snapshot(snapshot: SnapshotAssertion) -> None:
    """Test that project information is preserved in graph snapshots."""
    config = MetaxyConfig(project="snapshot_project")
    MetaxyConfig.set(config)

    try:
        graph = FeatureGraph()

        with graph.use():

            class Feature1(
                Feature,
                spec=FeatureSpec(
                    key=FeatureKey(["feature1"]),
                    fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
                ),
            ):
                pass

            class Feature2(
                Feature,
                spec=FeatureSpec(
                    key=FeatureKey(["feature2"]),
                    fields=[FieldSpec(key=FieldKey(["default"]), code_version=1)],
                ),
            ):
                pass

        # Get snapshot
        snapshot_data = graph.to_snapshot()

        # Both features should have project in snapshot
        # The snapshot format should include project information
        # (Implementation detail for python-dev agent)

        # For now, verify features have correct projects
        assert Feature1.project == "snapshot_project"
        assert Feature2.project == "snapshot_project"

        # Snapshot the graph structure
        assert {
            "feature1_project": Feature1.project,
            "feature2_project": Feature2.project,
            "feature_keys": list(snapshot_data.keys()),
        } == snapshot

    finally:
        MetaxyConfig.reset()


def test_data_version_by_field_unchanged_by_project(
    snapshot: SnapshotAssertion,
) -> None:
    """Test that data_version() method is also unchanged by project.

    This verifies the entire data versioning pipeline ignores project metadata.
    """
    graph_a = FeatureGraph()
    graph_b = FeatureGraph()

    config_a = MetaxyConfig(project="proj_a")
    MetaxyConfig.set(config_a)

    with graph_a.use():

        class FeatureA(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["multi", "field"]),
                fields=[
                    FieldSpec(key=FieldKey(["field1"]), code_version=1),
                    FieldSpec(key=FieldKey(["field2"]), code_version=2),
                ],
            ),
        ):
            pass

    config_b = MetaxyConfig(project="proj_b")
    MetaxyConfig.set(config_b)

    with graph_b.use():

        class FeatureB(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["multi", "field"]),
                fields=[
                    FieldSpec(key=FieldKey(["field1"]), code_version=1),
                    FieldSpec(key=FieldKey(["field2"]), code_version=2),
                ],
            ),
        ):
            pass

    # data_version() should be identical
    data_version_a = FeatureA.data_version()
    data_version_b = FeatureB.data_version()

    assert data_version_a == data_version_b
    assert data_version_a == snapshot

    # But projects should differ
    assert FeatureA.project != FeatureB.project

    MetaxyConfig.reset()
