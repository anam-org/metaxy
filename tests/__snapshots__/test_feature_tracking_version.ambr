# serializer version: 1
# name: test_data_version_by_field_unchanged_by_project
  dict({
    'field1': '9cb9f5bd9b380278e273033d16f07feabda7bdf5b813e9941b0f29e63b20efdb',
    'field2': '0b485f68befbb7abb029b43a698161745c63259b8ad0bc00260b7da4b9830243',
  })
# ---
# name: test_feature_tracking_version_includes_project
  dict({
    'feature_version_a': '3731d5aef6d53cfda612433171c52401406f94fd1c3482e7754ef2b24d85d170',
    'feature_version_b': '3731d5aef6d53cfda612433171c52401406f94fd1c3482e7754ef2b24d85d170',
    'feature_versions_same': True,
    'project_a': 'project_a',
    'project_b': 'project_b',
    'tracking_version_a': '7da02a89202f8eb9f19782512fadf988b815df8255e52a1f35dec9da76e18b12',
    'tracking_version_b': '5908b5019bdc1acd103482f1d144338f407f50d9d1aa09ca3808fe376f5ffd98',
    'tracking_versions_differ': True,
  })
# ---
# name: test_feature_version_unchanged_by_project
  'fcd6985a4dea45488450141be5c4712fd47781b4d0a7e42f04760c376eadf42f'
# ---
# name: test_feature_with_deps_different_projects
  dict({
    'downstream_a_project': 'project_a',
    'downstream_a_version': 'b7478f4be75eea0fa8d468df3f33cc8033cde080313e4ef783222fd65416f8e2',
    'downstream_b_project': 'project_b',
    'downstream_b_version': 'b7478f4be75eea0fa8d468df3f33cc8033cde080313e4ef783222fd65416f8e2',
    'upstream_a_project': 'project_a',
    'upstream_a_version': '8fd9019605c792a83a2cab5aaac4aea7a1e8524e51b728c0e02beb8032ee0ec1',
    'upstream_b_project': 'project_b',
    'upstream_b_version': '8fd9019605c792a83a2cab5aaac4aea7a1e8524e51b728c0e02beb8032ee0ec1',
  })
# ---
# name: test_project_in_graph_snapshot
  dict({
    'feature1_project': 'snapshot_project',
    'feature2_project': 'snapshot_project',
    'feature_keys': list([
      'feature1',
      'feature2',
    ]),
  })
# ---
# name: test_tracking_version_deterministic
  '139c7982886b761076245c878e6ad545f4afb9ffabab9238d7af883aff161b9b'
# ---
